src/lsm/cache_map.zig

```diff
@@ -0,0 +1,358 @@
+/// Should an insert() cause an eviction (which can happen either because the Key is the same,
+/// or because our Way is full), the evicted value is caught and put in the HashMap.
+///
+/// Cache invalidation for the HashMap is then handled by `compact`.
+pub fn CacheMap(
+    comptime Key: type,
+    comptime Value: type,
+    comptime key_from_value: fn (*const Value) callconv(.Inline) Key,
+    comptime hash_: fn (Key) callconv(.Inline) u64,
+    comptime equal: fn (Key, Key) callconv(.Inline) bool,
+    comptime tombstone_from_key: fn (Key) callconv(.Inline) Value,
+    comptime tombstone: fn (*const Value) callconv(.Inline) bool,
+) type {
+    const _Cache = SetAssociativeCache(
+        Key,
+        Value,
+        key_from_value,
+        hash_,
+        equal,
+        .{},
+    );
+
+    const HashMapContextValue = struct {
+        const Self = @This();
+
+        pub inline fn eql(_: Self, a: Value, b: Value) bool {
+            return equal(key_from_value(&a), key_from_value(&b));
+        }
+
+        pub inline fn hash(_: Self, value: Value) u64 {
+            return stdx.hash_inline(key_from_value(&value));
+        }
+    };
+
+    const load_factor = 50;
+    const _Map = std.HashMapUnmanaged(
+        Value,
+        void,
+        HashMapContextValue,
+        load_factor,
+    );
+
+    return struct {
+        const Self = @This();
+
+        pub const Cache = _Cache;
+        pub const Map = _Map;
+
+        pub const CacheMapOptions = struct {
+            cache_value_count_max: u32,
+            map_value_count_max: u32,
+            scope_value_count_max: u32,
+            name: []const u8,
+        };
+
+        // The hierarchy for lookups is cache -> map_1 -> map_2. Lower levels _may_ have stale
+        // values, provided the correct value exists in the level above.
+        cache: Cache,
+        map_1: Map,
+        map_2: Map,
+
+        // Scope support.
+        // 1. After an upsert, we have evicted an item that was an exact match. This means we're
+        //    doing an update of an item that's in the cache. Store the original item without
+        //    clobbering in our scope_map.
+        // 2. After an upsert, we haven't evicted anything, check our stash:
+        //    a. If a matching item exists there, it means we're doing an update of an item that's
+        //       in the stash. Store the original item without clobbering in our scope_map.
+        //    b. If no matching item exists there, it means it's an insert. Store a tombstone
+        //       without clobbering in our scope_map.
+        // 3. After an upsert, we have evicted an item that was not an exact match. This means we're
+        //    doing an insert of a new value, but two keys have the same tags. Store the evicted
+        //    item without clobbering in our scope_map
+        scope_is_active: bool = false,
+        scope_map: Map,
+
+        last_upsert_was_update_with_eviction: bool = undefined,
```

Comment:
> unless this is a very hot performance path

It is a hot path, but I'll check and see if there's much impact - agreed re UB.

Reply:
So `?bool = null` is consistently faster by a good ~4% on the benchmark on tmpfs :exploding_head: 

(if it'll get used or not will depend how it compares with linearizing the control flow...)

Reply:
Nice! Also surprising. :exploding_head: Especially since there's no runtime safety check for [branching on undefined values](https://github.com/ziglang/zig/issues/63).